### ISSUE TITLE
bug: `decode2text.sh` is missing (used for Dovecot FTS - Xapian)

### ISSUE THREAD

**Issue by cfis** [docker-mailserver/docker-mailserver#4229]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

After upgrading to the latest release (via the Helm chart), I now see this error message in the log:

```
dovecot: decode2text: Fatal: execvp(/usr/lib/dovecot/decode2text.sh) failed: No such file or director
```
I searched the container, but don't see decode2text.sh installed anyplace.  Am I missing it, or is it no longer included in the latest build?

If its location has changed, I'll need to update the helm chart:

https://github.com/docker-mailserver/docker-mailserver-helm/blob/0ad4d474ba28d1bd6200990446c4a89686d4a817/charts/docker-mailserver/values.yaml#L552

This impacts full text search (I have Xapian enabled)

### 🐋 DMS Version

14.0.0

### 💻 Operating System and Architecture

Fedora 40 

### ⚙️ Container configuration files

```yml
Using helm chart, running on Kubernetes
```

**Comment by casperklein** [comment]
The location seems to have changed to `/usr/share/doc/dovecot-core/examples/decode2text.sh`.

However, it's deleted on build:

https://github.com/docker-mailserver/docker-mailserver/blob/34eb54ac398c4c3846645cf79404d8528869891b/Dockerfile#L292

**Comment by polarathene** [comment]
The [current Dovecot CE docs](https://doc.dovecot.org/main/core/summaries/settings.html#fts_decoder) (_equivalent [plugins reference](https://doc.dovecot.org/main/core/plugins/fts.html#fts_decoder)_) suggest it should be at `/usr/lib/dovecot/decode2text.sh`.

So this is probably because of the change away from CE builds since v14 upgraded to Debian 12?

[Here is the official source](https://github.com/dovecot/core/blob/16de26fc328d188801092bfbedba43255cc7859f/src/plugins/fts/decode2text.sh) where it doesn't seem to have been changed for over 2 years. So you could also grab it from there 🤔 (_weirdly the [most recent tagged release `2.3.21.1`](https://github.com/dovecot/core/blob/2.3.21.1/src/plugins/fts/decode2text.sh) from Aug 2024 is a commit behind `main` branch, that commit is 12 years old_).

I wonder if it might be wiser for us to move back to Ubuntu? That would be fairly similar to Debian right? With a few differences to account for... oh it seems worse [Dovecot CE](https://repo.dovecot.org/) hasn't had support for Ubuntu since `20.04`, I guess the devs are really stubborn about not supporting newer distro releases until they get Dovecot 2.4 + 3.0 out 🤷‍♂️ 

Ubuntu 24.04 presently packages the [2nd latest `2.3.21`](https://github.com/dovecot/core/releases/tag/2.3.21) release (Sep 2023). Fedora 41 will be officially released in November and has the [current latest release `2.3.21.1`](https://github.com/dovecot/core/releases/tag/2.3.21.1) (Aug 2024).

---

I suppose while we still support Xapian we should add a test case to the test suite for catching this failure and we can temporarily fix it by copying the `decode2text.sh` to `/usr/lib/dovecot/decode2text.sh` where we'd expect it? (_ownership is `root:root` with standard `755` permissions_)

**Comment by casperklein** [comment]
> However, it's deleted on build:
> 
> https://github.com/docker-mailserver/docker-mailserver/blob/34eb54ac398c4c3846645cf79404d8528869891b/Dockerfile#L292

For reference, I did some more testing. Even when commenting out the line and building the image, the file is missing. That is because we use `debian:12-slim`. With the `slim` tag, `/usr/share/doc` is always empty.

**Comment by cfis** [comment]
Any decision on how to proceed?

**Comment by polarathene** [comment]
> Any decision on how to proceed?

Changing base image isn't likely any time soon. Dovecot CE builds are no longer available to us either until Dovecot devs decide to support Debian 12.

Since the file doesn't appear to change much, we could probably just bundle a direct copy of it in `target/dovecot/` and `COPY` to expected location in `Dockerfile`? Seems the easiest solution.

As noted, this is not the only file missing since the upgrade to Debian 12. Not sure if that matters, we could just assume that it's fine unless there's an issue raised again about missing files?

---

**TL;DR:** Just do a PR with [this `decode2text.sh` file](https://github.com/dovecot/core/blob/16de26fc328d188801092bfbedba43255cc7859f/src/plugins/fts/decode2text.sh) in our repo `target/dovecot/fts/decode2text.sh` and `COPY` to `/usr/lib/dovecot/decode2text.sh` in the `Dockerfile` 👍

**Comment by cfis** [comment]
Yup ok. Are you doing the PR or you want me to?

**Comment by polarathene** [comment]
I could see that it's done for v15 release which should be out before end of year, but if you'd like to get the PR up for me that's always appreciated 😅

**Comment by freman** [comment]
Unfortunately one cannot just add the shell script, it has other missing dependancies

```
decode2text: Error: /usr/local/bin/decode2text.sh: 70: /usr/local/bin/xml2text: not found
```

**Comment by polarathene** [comment]
I'm not seeing any `decode2text` in configs or scripts of our image. I'm only aware of this snippet in our docs:

https://github.com/docker-mailserver/docker-mailserver/blob/6b4627ceab44c65d04ae25926aab05956ca2bafb/docs/content/config/advanced/full-text-search.md?plain=1#L47-L53

That was contributed in 2021 as part of the FTS Xapian support: https://github.com/docker-mailserver/docker-mailserver/issues/2061

Anyone relying on this has chosen to configure this themselves? It's only required if you want to index attachments, since this is not a feature DMS has any official integration for opt-out is manual by the user that added such config. Alternatively consider a different indexer?

- If someone would like to PR the support for xapian to index attachments with `decode2text.sh` that is welcome, ideally with a test to ensure the functionality continues to work and is not lost.
- As flatcurve is meant to replace xapian, it might be worthwhile for such a user to investigate what attachment indexing is like for that if it does not use `decode2text.sh`, since AFAIK it's the intended successor?

**Comment by cfis** [comment]
@polarathene this code was present in versions before 14.0. Version 14.0 removed it (indirectly) due to switching to Debian 12 (due to how Dovecot is packaged). So what used to work (indexing emails) no longer works, broken simply by upgrading. Of course indexing is a pretty important feature when you have thousands of emails.

**Comment by polarathene** [comment]
> So what used to work (indexing emails) no longer works, broken simply by upgrading.

I understand, but I don't think we officially ever supported it. We have no scripts integration at the DMS level, only the docs example where this service has always been commented out.

```console
$ docker run --rm -it mailserver/docker-mailserver:13 bash
$ apt update && apt install ripgrep
$ rg decode2text /etc

$ rg decode2text /usr
/usr/lib/dovecot/decode2text.sh
10:#   fts_decoder = decode2text
12:# service decode2text {
13:#   executable = script /usr/local/libexec/dovecot/decode2text.sh
15:#   unix_listener decode2text {
```

You see nothing is enabled in the earlier DMS releases either. This was something users chose to opt-in, and it would have only worked for as long as we defaulted to the community repo of Dovecot which was v12 (April 2023):
- https://github.com/docker-mailserver/docker-mailserver/pull/2901
- https://github.com/docker-mailserver/docker-mailserver/releases/tag/v12.0.0

The ability to use it without a custom DMS build + the custom config from the user would have only been available for a little over a year before v14 was released.

---

If the helm chart has added something to officially support this, that is not an issue for our official DMS image in my opinion 🤔 

While I absolutely understand it's unfortunate for the users affected, since there's no clear sign of when Dovecot will release 2.4 + 3.0 (_they'll be the same release, but one is behind a paywall IIRC with some differences_), we don't know when the community repo will get Debian 12 support. AFAIK it's not just Debian 12 but any other distro the community repo has had support for. Dovecot doesn't follow semver, so [`2.4` will be a breaking change](https://doc.dovecot.org/main/installation/upgrade/2.3-to-2.4.html) to upgrade to once available.

If you can build the equivalent Dovecot community package without too much trouble, we could consider that. It'd probably need a separate repo that publishes the package though to avoid wasting our CI resources for a feature that is likely not used by many DMS users AFAIK. I'm open to switching the base image, but the issue would be present anywhere else too I think.

Someone will likely have to do more work to piece it together manually as described above. That will need to come from someone who cares about the feature to warrant that time. In the past I'd normally take care of it, but I've been at my capacity with a long backlog for quite some time now, so I can no longer commit to more tasks :(

Personally, I would encourage evaluating the situation with flatcurve before doing the work for xapian.

---

**EDIT:** flatcurve is xapian based, it'll be [bundled with Dovecot 2.4](https://doc.dovecot.org/main/core/plugins/fts_flatcurve.html#fts_flatcurve_index) out of the box instead of requiring us to build as a separate plugin AFAIK.

Not 100% sure if we can support it before Dovecot 2.4? But it looks like there's [two indexers for attachments supported by Dovecot](https://doc.dovecot.org/main/core/plugins/fts.html#indexing-attachments). One is the `decode2text` discussed here, another is Apache Tika. Similar to Solr, you could use a [separate docker image](https://hub.docker.com/r/apache/tika) and figure out how to configure/integrate that for DMS 👍 

Just noticed:
- Both Tika and Solr are Apache projects, so same Java concern vs Xapian there. Prior Dovecot series like 2.3.x had an FTS engine for Solr, whereas 2.4 seems to now be Tika? (**EDIT:** Nope, [there's still Solr support](https://doc.dovecot.org/main/core/summaries/settings.html#fts_solr) via the [`fts-solr` plugin](https://doc.dovecot.org/main/core/plugins/fts_solr.html)), Flatcurve support has it's own [`fts-flatcurve` plugin docs](https://doc.dovecot.org/main/core/plugins/fts_flatcurve.html) too.
- [Solr uses Tika for indexing binary files](https://solr.apache.org/guide/solr/latest/indexing-guide/indexing-with-tika.html) and [Solr itself apparently bundles Tika](https://www.reddit.com/r/Solr/comments/ill5ds/comment/g3tk5c9/) internally, whereas the image I linked is for the Tika server alone 🤷‍♂️ 

So that seems like it's Dovecot 2.4 specific, but the xapian based flatcurve plugin could use Tika for indexing attachments too?
